client_ext_test.go

```diff
@@ -106,10 +106,8 @@ func TestClientPeer(t *testing.T) {
 		err = clientStream.Send(&pingv1.SumRequest{})
 		assert.Nil(t, err)
 		// server streaming
-		serverStream, err := client.CountUp(ctx, connect.NewRequest(&pingv1.CountUpRequest{}))
+		serverStream, err := client.CountUp(ctx, connect.NewRequest(&pingv1.CountUpRequest{Number: 1}))
```

Comment:
It causes the server to error with `invalid_argument: number must be positive: got 0`

Reply:
The test is checking in the interceptor that the peer address and peer protocol is set. It's testing each request style by sending an empty but valid request, however server stream is the only one that will cause the server handler to error. This can now report the server error instead of success. I don't think this is intended behaviour of the test.

Reply:
Not setting `Number` will cause the following only for grpcweb:
```
--- FAIL: TestClientPeer (0.00s)
    --- FAIL: TestClientPeer/grpcweb (0.07s)
        client_ext_test.go:112:
            assertion:  assert.Nil
            got:        unknown: http2: response body closed
```
- `Close` to fail with [`http2: response body closed`](https://github.com/golang/net/blob/48a597731ce9fbe97fed1733da5e808233b6cdcb/http2/transport.go#L2603) [here](https://github.com/connectrpc/connect-go/blob/d9b61e93e4e7e8c2a4c6be7a31fee7c33e7dc9a3/client_ext_test.go#L114)
- Errors on [`discard`](https://github.com/connectrpc/connect-go/blob/23dc2dae9871464ff47521548c5a29f41596ef5b/duplex_http_call.go#L180) which usually receives an `io.EOF`

The grpcweb error response is a header only response, need to investigate why this error is returned on a no body.

Reply:
Now flaky with: 
```
            client_ext_test.go:113:
                assertion:      assert.Nil
                got:    unknown: io: read/write on closed pipe
```
Discard is reading from a closed response which is causing an ` io.ErrClosedPipe`. This is probably due to the new memhttp. Need to investigate.

Reply:
> Now flaky with:

By "now", do you mean after the last commit you pushed, which removed calls to `response.Body.Close()`?

Also, in that latest commit, who is the caller that is now responsible for calling `Close`? Is this done when `StreamClientConn.CloseResponse()` is called? Is there a chance some code path is _failing_ to call that? (In particular, I'm suspicious why this would only happen for gRPC-Web and not others.)

> Discard is reading from a closed response which is causing an `io.ErrClosedPipe`.

It seems like `discard` should be resilient against `io.ErrClosedPipe` and replace it with `io.EOF` when observed, no?

Reply:
It's due to the change in closing the pipe writer on Reader instead of Writer. `net.Pipe` will close both halves but the error is given on the opposite end. So closing the Reader made the Writer return nice `io.EOF` errors but the `Reader` to receive `io.ErrClosedPipe`. HTTP2 libraries set the request error to the response [here](https://github.com/golang/net/blob/4c7a5b64f145becd6fbd7d22f7ad4e9b891ce43a/http2/transport.go#L1326)

I'll need to write more testing around closing, and create an issue for testing with different latencies.

Reply:
HTTP2 sets the error on the response then [here](https://github.com/golang/net/blob/4c7a5b64f145becd6fbd7d22f7ad4e9b891ce43a/http2/transport.go#L1578)

Reply:
Discard should never see `io.ErrClosedPipe`, however I've added context error checks to avoid issues with cancelation errors bubbling up to `CloseResponse()` errors.

Working and passing with `-count=100` locally, can't see any more flaky test issues. Could you please re-review.